File: frontend/src/types/anomaly.ts
/**
 * TypeScript interfaces for Phase 15: Traffic Anomaly & Congestion Incident Detection.
 * Strictly typed with full provenance tracking, explainable rule triggers, and lifecycle states.
 */

export type AnomalyRuleType =
  | 'congestion_buildup'
  | 'flow_drop'
  | 'lane_imbalance'
  | 'density_spike';

export type AnomalySeverity = 'low' | 'medium' | 'high' | 'critical';

export type AnomalyStatus = 'open' | 'acknowledged' | 'resolved';

export interface AnomalyEvent {
  id: string;
  session_id: string;
  anomaly_type: AnomalyRuleType | string;
  severity: AnomalySeverity;
  status: AnomalyStatus;
  title: string;
  description: string;
  start_timestamp_seconds: number;
  end_timestamp_seconds?: number | null;
  duration_seconds: number;
  metric_name: string;
  trigger_value: number;
  baseline_value?: number | null;
  threshold_value: number;
  deviation_pct?: number | null;
  lane_id?: string | null;
  provenance_category: string;
  is_synthetic: boolean;
  details_json?: Record<string, any> | null;
  created_at: string;
  updated_at: string;
  resolved_at?: string | null;
}

export interface AnomalyEventListResponse {
  events: AnomalyEvent[];
  total: number;
  limit: number;
  offset: number;
  active_count: number;
  provenance_breakdown: Record<string, number>;
}

export interface AnomalyEventDetailResponse {
  event: AnomalyEvent;
  video_id?: string | null;
  video_filename?: string | null;
  video_source_type?: string | null;
  provenance_verified: boolean;
  source_reference?: string | null;
  license_reference?: string | null;
}

export interface AnomalyRuleConfigSchema {
  rule_name: string;
  anomaly_type: string;
  description: string;
  threshold_value: number;
  threshold_unit: string;
  condition_description: string;
  severity_criteria: Record<string, string>;
  caveats?: string | null;
}

export interface AnomalyInfoResponse {
  version: string;
  phase: string;
  description: string;
  terminology_disclaimer: string;
  anti_fabrication_policy: string;
  rules: AnomalyRuleConfigSchema[];
}

export interface DetectAnomaliesResponse {
  session_id: string;
  anomalies_detected: number;
  new_events_count: number;
  updated_events_count: number;
  events: AnomalyEvent[];
  execution_time_ms: number;
}

export interface UpdateAnomalyStatusRequest {
  status: AnomalyStatus;
  note?: string;
}

export interface AnomalyQueryParams {
  session_id?: string;
  anomaly_type?: string;
  severity?: string;
  status?: string;
  provenance_category?: string;
  is_synthetic?: boolean;
  limit?: number;
  offset?: number;
}

File: backend/app/schemas/anomaly.py
"""
Pydantic v2 schemas for Traffic Anomaly & Congestion Incident Detection.
Phase 15: Traffic Anomaly & Congestion Incident Detection.
"""
from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, ConfigDict, Field


class AnomalyEventSchema(BaseModel):
    """Schema representing a single persisted traffic anomaly / incident event."""
    model_config = ConfigDict(from_attributes=True)

    id: str
    session_id: str
    anomaly_type: str = Field(
        ...,
        description="Type of anomaly: 'congestion_buildup', 'abnormal_flow_drop', 'lane_imbalance', or 'density_spike'",
    )
    severity: str = Field(
        ...,
        description="Severity level: 'low', 'medium', 'high', or 'critical'",
    )
    status: str = Field(
        default="open",
        description="Lifecycle status: 'open', 'acknowledged', or 'resolved'",
    )
    title: str
    description: str
    start_timestamp_seconds: float
    end_timestamp_seconds: Optional[float] = None
    duration_seconds: float
    metric_name: str
    trigger_value: float
    baseline_value: Optional[float] = None
    threshold_value: float
    deviation_pct: Optional[float] = None
    lane_id: Optional[str] = None
    provenance_category: str = Field(
        default="real_database_metrics",
        description="Inherited provenance: 'real_database_metrics', 'synthetic_pipeline_metrics', 'synthetic_fixture', or 'unavailable'",
    )
    is_synthetic: bool = False
    details_json: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime
    resolved_at: Optional[datetime] = None


class AnomalyEventListResponse(BaseModel):
    """Paginated response containing filtered anomaly events."""
    model_config = ConfigDict(from_attributes=True)

    events: List[AnomalyEventSchema]
    total: int
    limit: int
    offset: int
    active_count: int
    provenance_breakdown: Dict[str, int] = Field(default_factory=dict)


class AnomalyEventDetailResponse(BaseModel):
    """Detailed response for a single anomaly event with session and video lineage."""
    model_config = ConfigDict(from_attributes=True)

    event: AnomalyEventSchema
    video_id: Optional[str] = None
    video_filename: Optional[str] = None
    video_source_type: Optional[str] = None
    provenance_verified: bool = False
    source_reference: Optional[str] = None
    license_reference: Optional[str] = None


class AnomalyRuleConfigSchema(BaseModel):
    """Metadata describing a specific anomaly detection rule and its active threshold."""
    model_config = ConfigDict(from_attributes=True)

    rule_name: str
    anomaly_type: str
    description: str
    threshold_value: float
    threshold_unit: str
    condition_description: str
    severity_criteria: Dict[str, str]
    caveats: Optional[str] = None


class AnomalyInfoResponse(BaseModel):
    """Metadata describing Phase 15 detection capabilities, rule configurations, and safety disclaimers."""
    model_config = ConfigDict(from_attributes=True)

    version: str = "1.0.0"
    phase: str = "Phase 15 — Traffic Anomaly & Congestion Incident Detection"
    description: str = (
        "Statistical and rule-based anomaly detection engine evaluating persisted traffic "
        "flow, lane occupancy, density spikes, and directional distributions."
    )
    terminology_disclaimer: str = (
        "Anomaly = statistical deviation; Incident = persisted threshold event. "
        "Does NOT imply vehicle collision, accident, or emergency dispatch event."
    )
    anti_fabrication_policy: str = (
        "Anomalies are detected strictly from genuine metric threshold crossings. "
        "Zero synthetic fabrication: if no anomaly condition exists, zero events are created."
    )
    rules: List[AnomalyRuleConfigSchema]


class UpdateAnomalyStatusRequest(BaseModel):
    """Request payload to acknowledge or resolve an anomaly event."""
    model_config = ConfigDict(extra="ignore")

    status: str = Field(
        ...,
        description="Target status: 'acknowledged', 'resolved', or 'open'",
    )
    note: Optional[str] = Field(
        default=None,
        max_length=512,
        description="Optional operator note regarding status update",
    )


class DetectAnomaliesResponse(BaseModel):
    """Response returned when on-demand anomaly detection is executed for a session."""
    model_config = ConfigDict(from_attributes=True)

    session_id: str
    anomalies_detected: int
    new_events_count: int
    updated_events_count: int
    events: List[AnomalyEventSchema]
    execution_time_ms: float
